opal/mca/common/ofi/common_ofi.c

```diff
@@ -25,3 +26,291 @@ int mca_common_ofi_register_mca_variables(void)
+    }
+
+    return logical_index;
+}
+
+/* Takes a hardware topology and finds the NUMA node that the current process is on.
+ * If the process is not bound, returns OPAL_ERROR
+ */
+static int
+get_process_numanode(hwloc_topology_t topology)
+{
+    int ret, logical_index = OPAL_ERROR;
+    hwloc_bitmap_t cpuset;
+    hwloc_obj_t obj = NULL;
+
+    if (NULL == topology) {
+        return OPAL_ERROR;
+    }
+
+    /* allocate memory for cpuset */
+    cpuset = hwloc_bitmap_alloc();
+    if (NULL == cpuset) {
+        return OPAL_ERROR;
+    }
+
+    /* fill cpuset with the collection of cpu cores that the process runs on */
+    ret = hwloc_get_cpubind(topology, cpuset, HWLOC_CPUBIND_PROCESS);
+    if (0 > ret) {
+        goto error;
+    }
+
+    /* get the package that contains the cpuset */
+    obj = hwloc_get_first_largest_obj_inside_cpuset(topology, cpuset);
+    if (NULL == obj) {
+        goto error;
+    }
+
+    /* return the NUMA node's logical index under this package */
+    if (NULL != obj-> memory_first_child) {
+        logical_index = obj->memory_first_child->logical_index;
+    }
+
+error:
+    hwloc_bitmap_free(cpuset);
+    return logical_index;
+}
+
+/* Count providers returns the number of providers present in an fi_info list
+ *     @param (IN) provider_list    struct fi_info* list of providers available
+ *
+ *     @param (OUT)                 int number of providers present in the list
+ *
+ *     returns 0 if the list is NULL
+ */
+static int
+count_providers(struct fi_info* provider_list)
+{
+    struct fi_info* dev = provider_list;
+    int num_provider = 0;
+
+    while (NULL != dev) {
+        num_provider++;
+        dev = dev->next;
+    }
+
+    return num_provider;
+}
+
+/* Selects a NIC based on hardware locality to process NUMA node and device BDF.
+ *
+ * inputs:
+ *     @param provider_list (IN)    struct fi_info* to an initially selected
+ *                                  provider NIC. The provider name and
+ *                                  attributes are used to restrict NIC
+ *                                  selection. This provider is returned if the
+ *                                  NIC selection fails.
+ *
+ *     @param proc_numa_index (IN)  int Current numa rank. used for load
+ *                                  balancing if multiple valid providers are
+ *                                  found under the same NUMA node, or no providers
+ *                                  are found under the same NUMA node but multiple
+ *                                  valid providers are available.
+ *
+ * returns:
+ *     @param provider (OUT)        struct fi_info* object with the selected
+ *                                  provider if the selection succeeds
+ *                                  if the selection fails, returns the fi_info
+ *                                  object that was initially provided.
+ *
+ * If there is more than one provider that shares the same NUMA node, we use
```

Comment:
What's the rationale behind using this approach?

Reply:
My general goal was to distribute the processes across all available NIC as much as possible while choosing a local NIC when available. Is there a specific aspect that you wanted me to comment on?